crates/iceberg-catalog/src/service/storage/s3.rs

```diff
@@ -356,6 +423,82 @@ impl S3Profile {
         Ok(config)
     }
 
+    async fn get_aws_sts_token(
+        &self,
+        table_location: &str,
+        cred: &S3Credential,
+        arn: &String,
+    ) -> Result<aws_sdk_sts::types::Credentials, TableConfigError> {
+        self.get_sts_token(table_location, cred, Some(arn)).await
+    }
+
+    async fn get_minio_sts_token(
+        &self,
+        table_location: &str,
+        cred: &S3Credential,
+    ) -> Result<aws_sdk_sts::types::Credentials, TableConfigError> {
+        self.get_sts_token(table_location, cred, None).await
+    }
+
+    async fn get_sts_token(
+        &self,
+        table_location: &str,
+        cred: &S3Credential,
+        arn: Option<&str>,
+    ) -> Result<aws_sdk_sts::types::Credentials, TableConfigError> {
+        let cred = self
+            .get_config(self.get_aws_sdk_credentials(Some(cred))?)
+            .await;
+
+        let assume_role_builder = aws_sdk_sts::Client::new(&cred)
+            .assume_role()
+            .role_session_name("iceberg")
+            .policy(Self::get_aws_policy_string(table_location));
+        let assume_role_builder = if let Some(arn) = arn {
+            assume_role_builder.role_arn(arn)
+        } else {
+            assume_role_builder
+        };
+
+        let v = assume_role_builder.send().await.map_err(|e| {
+            TableConfigError::FailedDependency(format!(
+                "aws::sts::assume_role token call failed: {e}"
+            ))
+        })?;
+        v.credentials.ok_or(TableConfigError::FailedDependency(
+            "aws::sts::assume_role token call response didn't contain credentials".to_string(),
+        ))
+    }
+
+    fn get_aws_policy_string(table_location: &str) -> String {
+        let resource = format!(
+            "arn:aws:s3:::{}",
```

Comment:
Why does this work? Don't we need a "*" at the end?

Reply:
A full ARN would be: arn:partition:service:region:account-id:resource-type:resource-id
For a managed catalog setup, we would assume an arn in a different account. The catalog is in account A, the client who owns the S3 bucket is in account B. Thus we would have to add at least account-id. I think adding the partition also wouldn't hurt. I don't think the region is necessary, but we should double check.

Probably we also need the the account-id then for assume arn above.

Reply:
Not sure I fully understand - we POST the client creds for the account owning the s3 bucket as the storage profile, anything happening inside here is then using those client creds, hence bucket & role etc are from that bucket-woning single account?